src/Worker/Grpc/GrpcOrchestrationRunner.cs

```diff
@@ -157,88 +157,97 @@ public static string LoadAndRun(
-            {
-                OrchestrationRuntimeState runtimeState = extendedSessionState!.RuntimeState;
-                runtimeState.NewEvents.Clear();
-                foreach (HistoryEvent newEvent in newEvents)
-                {
-                    runtimeState.AddEvent(newEvent);
-                }
-
-                result = extendedSessionState.OrchestrationExecutor.ExecuteNewEvents();
-                if (extendedSessionState.OrchestrationExecutor.IsCompleted)
-                {
-                    extendedSessions.Remove(request.InstanceId);
-                }
-            }
-            else
+            if (!pastEventsIncluded && extendedSessions.TryGetValue(request.InstanceId, out ExtendedSessionState? extendedSessionState) && extendedSessionState is not null)
             {
-                extendedSessions.Remove(request.InstanceId);
-                addToExtendedSessions = true;
+                OrchestrationRuntimeState runtimeState = extendedSessionState!.RuntimeState;
+                runtimeState.NewEvents.Clear();
+                foreach (HistoryEvent newEvent in newEvents)
+                {
+                    runtimeState.AddEvent(newEvent);
+                }
+
+                result = extendedSessionState.OrchestrationExecutor.ExecuteNewEvents();
+                if (extendedSessionState.OrchestrationExecutor.IsCompleted)
+                {
+                    extendedSessions.Remove(request.InstanceId);
+                }
+            }
+            else
+            {
+                extendedSessions.Remove(request.InstanceId);
+                addToExtendedSessions = true;
+            }
+        }
+
+        if (result == null)
+        {
+            // DurableTask.Core did not attach the orchestration history since the extended session is still active on its end, but we have since evicted the
+            // session and lost the orchestration history so we cannot replay the orchestration.
+            if (!pastEventsIncluded)
+            {
+                requiresHistory = true;
+            }
+            else
+            {
+                // Re-construct the orchestration state from the history.
+                // New events must be added using the AddEvent method.
+                OrchestrationRuntimeState runtimeState = new(pastEvents);
+
+                foreach (HistoryEvent newEvent in newEvents)
+                {
+                    runtimeState.AddEvent(newEvent);
+                }
+
+                TaskName orchestratorName = new(runtimeState.Name);
+                ParentOrchestrationInstance? parent = runtimeState.ParentInstance is ParentInstance p
+                    ? new(new(p.Name), p.OrchestrationInstance.InstanceId)
+                    : null;
+
+                DurableTaskShimFactory factory = services is null

```

Comment:
I only updated line 211 to 226.. the whole file changed because of the VS outlining thing